src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectWriter.cs

```diff
@@ -984,7 +984,7 @@ private long GetAssemblyId(WriteObjectInfo objectInfo)
             {
                 assemId = 0;
             }
-            else if (assemblyString.Equals(Converter.s_urtAssemblyString))
+            else if (Converter.s_urtAssembly.GetType(objectInfo.GetTypeFullName(), throwOnError: false) == objectInfo._objectType)
```

Comment:
I guess it comes with performance regressions. If you look at the discussion with eric above (is collapsed now) then you see that we were searching for other solutions. We don't really need this change if we can typeforward every serializable type which lives in mscorlib on netfx to mscorlib.
I haven't yet the found the code where primitive type arrays are defined. They aren't yet typeforwarded... Array itself is marked as TypeForwardedFrom but not the primitive implementations.